### ISSUE TITLE
Testing plan for website

### ISSUE THREAD

**Issue by alexwolfe** [gravitational/next#812]
We’re going to need better testing on the website when we make changes that effect a larger number of pages. Our site needs to be very reliable since it is a critical part of our business operations (including sales). 

## Testing should include 2 plans: 
- Basic testing plan for small changes effecting an isolated number of pages
- Full testing plan for large changes that effect all pages 

### Include in full testing plan
- A list of critical pages/sections to tests when global changes are made
- Forms that need to be tested
- A list of rarely used components that could cause issues (plus location) 
- Automated testing options we may be able to introduce (could include SEMRUSH) Ideally this would not rely on us manually writing many tests. 
- Other suggestions

**Comment by iAdramelk** [comment]
@alexwolfe started writing test plans and it is strongly correlating with the PR types. Made a list and some examples. Can you please check it and comment of the format, level of details, etc. Is what you wanted? Do I go in too much/too little details? Do I missed something? I will update the list and add rest of the items based on the feedback:

--

# Testing plan depending on the PR type

- New Component PR.
- Updating existing component PR.
- New Page PR.
- Updating existing page PR.
- Dependency update PR.
- New plugin/helper PR.
- Plugin/helper update PR.
- Full site test plan.

## New Component PR

If we want small PRs that are easy to review, it is a good idea to move each new component to separate PR.

PR content:

- Purpose of the component and how it will be used.
- Link to the design file or issue with it.
- Screenshots or video. Should cover states in the list below.
- Link to the live demo if available.
- Component API description (maybe a good idea discuss props format in the common chat before starting implementation, at lest in some non-trivial cases).
- Things to pay attention while reviewing code.
- Q: Do we need some separate rules to test internal logic of the components that process and convert data in some way or do we test them as a black box?

What to check at the code review:

- Component API (did it match task, does the props names are easy to understand, can we simplify it).
- Do the component do the stuff it should be.
- Do we understand what different parts of the code do and why? If not, we probably want to rename stuff or add comments. Good rule for comments - comment not what the code do, but why we need it to be done in the first place and why we do it is this way and not in the another.

To test how component works it is important to be able to run it in the browser and to run it in different states:

- Short texts.
- Long texts (preferably without spaces to check how we trim content).
- Optional fields filled.
- Optional fields not filled.
- Every combination of the above.
- Different modes (colours, sizes, etc).

After demo is launched we need to check for each example:

- Did the component matches design.
- How example working on at least these screen width: Mobile, 375x812 (mobile), 1024x768 (tablet), 1440x900 (laptop).
- Try to use component from the real phone if possible.
- Do interactive elements have `hover`, `active`, `focus` styles.
- Do interactive elements are accessible from the keyboard (switching elements by tab, closing popups by Esc, buttons pressed by enter/space, etc).
- Q: Do we check accessibility?
- Q: What browsers we should test it in?

Q: It is better if checking this demo does not require cloning repo locally and setting up all examples manually. Right now preparing demo of all states and displaying it can e done with storybook, if we remove storybook we need some other way to do it.

## Updating existing component PR

All point from the previous list applies here too to test for regressions, but we also need to test how the changes are affection pages and other components that uses this component.

PR content:

- Purpose of the change and affected areas.
Screenshots or video. Should cover states in the list below.
- Link to the live demo if available.
- If possible, list of components that uses this one as dependency. (Can be found by running the search on the repo)
- If possible list of places where these components are used on the pages directly. (Can be found by running the search on the repo)

What to check at the code review:

- Same as above.

Demos:

- Same as above.
- We also need to manually check other components' demos and affected pages that use this component. Demos for other components in different states can help a lot.

Q: Automation. We can run chromatic on the storybook to create and compare screenshots of every story for each PR. It should considerably simplify looking for regressions in other components that use the current one.

## New Page PR

I assume here that all separate component are already tested and merged, so we only looking on the page that uses them.

PR content:

- Purpose of the component and how it will be used.
- Link to the design file or issue with it.
- Screenshots or video. Should cover states in the list below.
- Link to the live demo.
- Things to pay attention while reviewing code.
- Interactive stuff we need to test (forms, redirects, etc).

What to check at the code review:

- Code and style specific of this page: Do we understand what different parts of the code do and why? If not, we probably want to rename stuff or add comments.
- I assume that at this stage all components are already merged and tested, so we can mostly skip discussing them, but feel free to write if you see what components and props are misused.
- Page metadata (do we have all the things we need or do we miss something).

After demo is launched we need to check for each example:

- Did the page matches design.
- How page working on at least these screen width: Mobile, 375x812 (mobile), 1024x768 (tablet), 1440x900 (laptop).
- Try to use page from the real phone if possible.
- Is there conflict between components.
- Do the navigation from keyboard make sense? (Order of switching between elements is correct, we can fill and submit forms, without using mouse, etc)
- Run lighthouse audit on the page and check the problems.
- Check for console errors.
- Check metadata (title, description, og:image, etc). Try to insert the link into the Facebook or into the [metadata tester](https://searchenginereports.net/open-graph-checker).

Q: We may add some important pages to the chromatic, but adding all pages to it would be overkill.


## Updating existing page PR

Use the list above and adopt it based on the changes. It may be not necessary to retest everything if we only updated line of text. But it is still a good idea to test updated line in different resolutions. 

To check for regressions it is a good idea to have current page opened in the next tab and to switch between then.

## Dependency update PR

TBD

## New plugin/helper PR

TBD

## Plugin/helper update PR

TBD

**Comment by alexwolfe** [comment]
Thanks this is a good start. I'm going to use parts of this to create a process for doing a proper pull request. My approach will be a progressive one meaning we'll work to improve the existing process but understand that over time we'll continue to refine the PR process as needed. One consideration is I don't want it to be cumbersome as it will cause us to move very slowly (or just not do it). Instead, I'd like to include what is essential and build from there. 

Thanks for your thoughts behind this, very helpful Alexey. Going to close this ticket for now as I have plenty of information here to create the first template for submitting a proper PR which we will review collectively as a team.
